graph-builder/src/registry.rs

```diff
@@ -37,23 +45,109 @@ impl Into<cincinnati::Release> for Release {
+    host: &str,
+    image: &str,
+    credentials_file: &str,
+) -> Result<RegistryClient, Error> {
+    let mut registry_client = dkregistry::v2::Client::configure(&tcore.handle())
+        .registry(host)
+        .insecure_registry(false)
+        .read_credentials(get_credentials_or_empty_reader(credentials_file)?)
+        .build()
+        .map_err(|e| format_err!("{}", e))?;
+
+    let fut_check_v2_prelogin = registry_client
+        .is_v2_supported()
+        .map_err(|e| format_err!("{}", e))?;
+    if !tcore
+        .run(fut_check_v2_prelogin)
+        .map_err(|e| format_err!("{}", e))?
+    {
+        return Err(format_err!("API v2 not supported by {}", host));
+    };
+
+    let fut_check_auth_prelogin = registry_client
+        .is_auth(None)
+        .map_err(|e| format_err!("{}", e))?;
+    let logged_in = tcore
+        .run(fut_check_auth_prelogin)
+        .map_err(|e| format_err!("{}", e))?;
+    if logged_in {
+        return Err(format_err!("no login performed, but already authenticated"));
+    }
+
+    let fut_login = registry_client
+        .login(vec![&format!("repository:{}:pull", image)])
+        .map_err(|e| format_err!("{}", e))?;
+    let token = tcore.run(fut_login).map_err(|e| format_err!("{}", e))?;
+
+    let fut_auth = registry_client
+        .is_auth(Some(token.token()))
+        .map_err(|e| format_err!("{}", e))?;
+    if !tcore.run(fut_auth).map_err(|e| format_err!("{}", e))? {
+        return Err(format_err!("login failed"));
+    }
+    registry_client.set_token(Some(token.token()));
+    debug!("successfully logged in to {}", host);
+
+    let fut_check_v2_postlogin = registry_client
+        .is_v2_supported()
+        .map_err(|e| format_err!("{}", e))?;
+    if !tcore
+        .run(fut_check_v2_postlogin)
+        .map_err(|e| format_err!("{}", e))?
+    {
+        return Err(format_err!("API v2 check failed after login to {}", host));
+    };
+
+    Ok(registry_client)
+}
+
 /// Fetches a vector of all release metadata from the given repository, hosted on the given
 /// registry.
-pub fn fetch_releases(registry: &str, repo: &str) -> Result<Vec<Release>, Error> {
+pub fn fetch_releases(
+    registry: &str,
+    repo: &str,
+    credentials_file: &str,
+) -> Result<Vec<Release>, Error> {
     let mut metadata = Vec::new();
-    for tag in fetch_tags(registry, repo)? {
-        metadata.push(Release {
-            source: format!(
-                "{}/{}:{}",
-                registry
-                    .trim_left_matches("https://")
-                    .trim_left_matches("http://"),
-                repo,
-                tag
-            ),
-            metadata: fetch_metadata(registry, repo, &tag)?,
-        })
+    let mut tcore = Core::new()?;
+
+    let registry_host = registry_trim(registry);
+    let registry_client =
+        get_authenticated_client(&mut tcore, registry_host, repo, credentials_file)?;
+
+    for tag in fetch_tags(&mut tcore, &registry_client, repo)? {
```

Comment:
Try something like this instead:

```rust
  fetch_tags(&mut tcore, &registry_client, repo)?
    .iter()
    .map(|tag| Release { source: ..., metadata: ... })
    .for_each(|release| {
      trace!("Found release {:?}", release);
      metadata.push(release);
    })
```

Reply:
Are you suggesting that for idiomatic reasons?

Reply:
Yep. Alternatively, you can do:

```rust
  for release in fetch_tags(&mut tcore, &registry_client, repo)?
    .iter()
    .map(|tag| Release { source: ..., metadata: ... })
  {
    trace!("Found release {:?}", release);
    metadata.push(release);
  }
```

That might be easier to follow.